### PR TITLE
Simplifiy tests infrastructure

### PR DESCRIPTION
**Pull request by giordano**
* remove the `runtests` function
* remove the `nfail` variable cruft
* generate a testset for each included file
* do a dummy `@test true` for each test in the loop: if the process errors out
the testset would error out as well

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -14,33 +14,24 @@ end
...
+            withenv("JULIA_NUM_THREAD" => "4") do
+                run(`$cmd -n $nprocs $(Base.julia_cmd()) $(joinpath(testdir, f))`; wait = true)
             end
+        else
+            run(`$cmd -n $nprocs $(Base.julia_cmd()) $(joinpath(testdir, f))`; wait = true)
         end
-        Base.with_output_color(:green,stdout) do io
-            println(io,"\tSUCCESS: $f")
-        end
+        wait(p)
```
Comment (simonbyrne) [inline test/runtests.jl:34]: isn't this redundant with `wait=true`?

File: test/runtests.jl [hunk 2]
```diff
@@ -14,33 +14,24 @@ end
...
-            cmd = `$cmd $args`
-            if f == "test_spawn.jl"
-                run(`$cmd -n 1 $(Base.julia_cmd()) $(joinpath(testdir, f))`)
-            elseif f == "test_threads.jl"
-                withenv("JULIA_NUM_THREAD" => "4") do
-                    run(`$cmd -n $nprocs $(Base.julia_cmd()) $(joinpath(testdir, f))`)
-                end
-            else
-                run(`$cmd -n $nprocs $(Base.julia_cmd()) $(joinpath(testdir, f))`)
+@testset "$f" for f in testfiles
```
Comment (simonbyrne) [inline test/runtests.jl:22]: is the `@testset` worth it? each one only has one test.

File: test/runtests.jl [hunk 3]
```diff
@@ -14,33 +14,24 @@ end
...
+                run(`$cmd -n $nprocs $(Base.julia_cmd()) $(joinpath(testdir, f))`; wait = true)
             end
+        else
+            run(`$cmd -n $nprocs $(Base.julia_cmd()) $(joinpath(testdir, f))`; wait = true)
         end
-        Base.with_output_color(:green,stdout) do io
-            println(io,"\tSUCCESS: $f")
-        end
+        wait(p)
+        @test success(p)
```
Comment (giordano) [inline test/runtests.jl:35]: The process `p` is run asynchronously, which has the side effect of making harder to see why it's failing.  As an (probably better) alternative, the process can be run with `wait = true` and here we can simply have
```suggestion
        @test true
```
If the process fails, `run` will throw an error and the tests will stop anyway

File: test/runtests.jl (hunk 1 above)
Comment (giordano) [reply test/runtests.jl:34]: In the REPL ``run(`...`; wait=true)`` returns immediately, but in a script it does seem to wait without this.  But see my comment below about not running the processes asynchronously

File: test/runtests.jl (hunk 2 above)
Comment (giordano) [reply test/runtests.jl:22]: It's just a way to get the names of the tests printed in a consistent fashion using the existing testing framework

File: test/runtests.jl (hunk 1 above)
Comment (simonbyrne) [reply test/runtests.jl:34]: > In the REPL `` run(`...`; wait=true) `` returns immediately

I can't replicate this:
```
julia> @time run(`sleep 3`; wait=true)
  3.003904 seconds (50 allocations: 1.906 KiB)
Process(`sleep 3`, ProcessExited(0))
```

File: test/runtests.jl (hunk 2 above)
Comment (simonbyrne) [reply test/runtests.jl:22]: fair enough.

**Comment by simonbyrne** [comment]
I'm not sure I understand the `wait=true` problem.

This now has conflicts after #390.

File: test/runtests.jl (hunk 1 above)
Comment (giordano) [reply test/runtests.jl:34]: I looked at my REPL history: I was testing
```julia
run(`sleep 10`; wait = false)
```
:expressionless:

**Comment by giordano** [comment]
Rebased on `master`.  I've just left the processes synchronous.

**Comment by simonbyrne** [comment]
Thanks!

**Pull request merged**